osu.Game/Rulesets/Difficulty/Skills/Skill.cs

```diff
@@ -1,120 +1,46 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
-using System;
 using System.Collections.Generic;
-using System.Linq;
 using osu.Game.Rulesets.Difficulty.Preprocessing;
-using osu.Game.Rulesets.Difficulty.Utils;
+using osu.Game.Rulesets.Mods;
 
 namespace osu.Game.Rulesets.Difficulty.Skills
 {
     /// <summary>
-    /// Used to processes strain values of <see cref="DifficultyHitObject"/>s, keep track of strain levels caused by the processed objects
-    /// and to calculate a final difficulty value representing the difficulty of hitting all the processed objects.
+    /// A bare minimal abstract skill for fully custom skill implementations.
     /// </summary>
     public abstract class Skill
     {
         /// <summary>
-        /// The peak strain for each <see cref="DifficultyCalculator.SectionLength"/> section of the beatmap.
+        /// <see cref="DifficultyHitObject"/>s that were processed previously. They can affect the difficulty values of the following objects.
         /// </summary>
-        public IReadOnlyList<double> StrainPeaks => strainPeaks;
+        protected readonly Stack<DifficultyHitObject> Previous = new Stack<DifficultyHitObject>(); // Contained objects not used yet
 
         /// <summary>
-        /// Strain values are multiplied by this number for the given skill. Used to balance the value of different skills between each other.
+        /// Visual mods for use in skill calculations.
         /// </summary>
-        protected abstract double SkillMultiplier { get; }
+        protected IReadOnlyList<ModWithVisibilityAdjustment> VisualMods => visualMods;
 
-        /// <summary>
-        /// Determines how quickly strain decays for the given skill.
-        /// For example a value of 0.15 indicates that strain decays to 15% of its original value in one second.
-        /// </summary>
-        protected abstract double StrainDecayBase { get; }
-
-        /// <summary>
-        /// The weight by which each strain value decays.
-        /// </summary>
-        protected virtual double DecayWeight => 0.9;
-
-        /// <summary>
-        /// <see cref="DifficultyHitObject"/>s that were processed previously. They can affect the strain values of the following objects.
-        /// </summary>
-        protected readonly LimitedCapacityStack<DifficultyHitObject> Previous = new LimitedCapacityStack<DifficultyHitObject>(2); // Contained objects not used yet
```

Comment:
not sure if we want to remove this as opposed to making the limited size configurable. probably need input from @smoogipoo on the reasoning / direction here (it may be done to limit retention of `DifficultyHitObjects`).

Reply:
I'm fairly certain this only exists in the codebase as a remnant of the original pp+ implementation for angle calculations.
Instead of having a configurable fixed size, what do we think about a method that allows the `Skill` to define before processing each object, which items are no longer needed?
Example use case: a reading skill that wants to have access to as many objects as are on the screen at a given time.

Reply:
If performance is our chief concern here, we can perhaps achieve the best of both worlds by having the `DifficultyCalculator` track the previous hitobject stack and simply pass the reference to it in the Skill ctors?

Reply:
I've addressed this in 13c12c246e6c7463626aec3d957349e32cdbe77b and 3a6aaf49f814afe1dd0f4703d662f99dcc849a3e.
13c12c246e6c7463626aec3d957349e32cdbe77b replaces LimitedCapacityStack<T> with DropoutStack<T>, which allows 3a6aaf49f814afe1dd0f4703d662f99dcc849a3e to implement a programmable history retention, rather than simply static size, or everything.
With these changes, we can have the stack retain only what is absolutely necessary as determined by the Skill, be that only the last n objects, objects within the last n milliseconds, or any other implementation skills could require.